Installer as a system container

giuseppe opened the pull request:
Allow to run the containerized installation using a system container instead of a Docker container.

A system container is managed by systemd and runs directly the container using runC instead of requiring another daemon as in the Docker case.

The advantage is that the installer can run on one of the hosts that is going to be used for the cluster.  Since the container doesn't run inside Docker, the installer can change the Docker configuration, restart its service without interfering with the installer container itself.

Work tracked here: https://trello.com/c/6zRfPN0Y/192-run-containerized-installer-as-system-container

giuseppe commented:
@ashcrow

ashcrow commented:
[test]

giuseppe commented:
[test]

giuseppe commented:
[test]

ashcrow commented:
aos-ci-test

ashcrow commented:
@cooktheryan if you have some time we'd appreciate a test of using this system container installer patch from the ops perspective.

ashcrow commented:
@sdodson we believe this is ready for code review.

cooktheryan commented:
@giuseppe and @ashcrow can you provide details on how to run this?

ashcrow commented:
@cooktheryan I think [this](https://github.com/openshift/openshift-ansible/pull/3720/files#diff-4f0b53ef4adc114f419dbfd541f5a538R18) should work. You'll also need to see the BUILD.md for the system container.

@giuseppe anything else you can think to add on this?

giuseppe commented:
@cooktheryan @ashcrow yes that should work.  Please build the container from `Dockerfile.rhel7` as this is the version we are going to use.
Also the log file can be configured with `--set VAR_LOG_OPENSHIFT_LOG=/file`, the default is `/var/log/ansible.log`

ashcrow commented:
aos-ci-test

ashcrow commented:
[test]

ashcrow commented:
@giuseppe I think a rebase to get past the following is needed:

```

[Fail] deploymentconfigs with revision history limits [Conformance] [It] should never persist more old deployments than acceptable after being observed by the controller 
/go/src/github.com/openshift/origin/test/extended/deployments/deployments.go:774
```

ashcrow commented:
Hrm, magic. It's all green now.

tbielawa commented on system-container/tmpfiles.template:
```diff
@@ -0,0 +1,2 @@
+d    $VAR_LIB_OPENSHIFT_INSTALLER - - - - -
```
What does this file do?

ashcrow replied:
That is the template configuration file for systemd-tmpfiles.

Variable is defined in https://github.com/openshift/openshift-ansible/pull/3720/files#diff-0152a7b8552088b3d310df6cfb5d6b76R133

tbielawa commented:
OK. It looks very strange to me. Like, "whoops, I accidentally committed a
temp garbage file".

Carry on.

On Mon, Apr 3, 2017 at 12:02 PM, Stephen Milner <notifications@github.com>
wrote:

> *@ashcrow* commented on this pull request.
> ------------------------------
>
> In system-container/tmpfiles.template
> <https://github.com/openshift/openshift-ansible/pull/3720#discussion_r109498837>
> :
>
> > @@ -0,0 +1,2 @@
> +d    $VAR_LIB_OPENSHIFT_INSTALLER - - - - -
>
> That is the template configuration file for systemd-tmpfiles.
>
> Variable is defined in https://github.com/openshift/
> openshift-ansible/pull/3720/files#diff-0152a7b8552088b3d310df6cfb5d6b
> 76R133
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/openshift-ansible/pull/3720#discussion_r109498837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACBYToNF5EKZ63frzVZgRfuXrF3r4jhks5rsUJAgaJpZM4Mj4OQ>
> .
>



-- 
Tim Bielawa, Software Engineer [ED-C137]
Cell: 919.332.6411  | IRC: tbielawa (#openshift)
1BA0 4FAB 4C13 FBA0 A036  4958 AD05 E75E 0333 AE37

giuseppe commented:
one know issue with this PR is that we support single files inventory files.  We may need to add a bind mount to the current directory to support multiple inventory files.

ashcrow commented:
I did a quick check looking at the Travis CI failures and it looks unrelated. The unit test failres are present on master. How they merged is anyones guess. 😕

codificat commented on README_ANSIBLE_CONTAINER.md:
```diff
@@ -0,0 +1,29 @@
+# Running ansible in a docker container
```
This file doesn't exist anymore, it was removed in commit 7fd854, the information about the container image is now in [README_CONTAINER_IMAGE.md](/openshift/openshift-ansible/blob/master/README_CONTAINER_IMAGE.md) and we should update that one instead as needed.

giuseppe replied:
Thanks.  I've fixed this, going to push a new version

codificat commented on system-container/run-system-container.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+export ANSIBLE_LOG_PATH=/var/log/ansible.log
+exec ansible-playbook -i /etc/ansible/hosts ${OPTS} ${PLAYBOOK_FILE}
```
The current image is based in playbook2image and it already provides a wrapper around `ansible-playbook` (the `run` script in the CMD) that handles things like running as non-root and more flexible inventory handling. Do we need a different command here? (I'm curious why...)

giuseppe replied:
We've added this part before `commit 18632f48e7c0ebe53ff1ec5bd894b3b198f769f2` so the two Dockerfiles were different.

As you pointed out, the best solution would probably be to add system containers support to playbook2image.

codificat replied:
`/etc/passwd` is group-writable by GID 0, and that trick allows `/usr/libexec/s2i/run` to work... it's still unclear to me why we need a different entry point, or if we need changes/fixes in playbook2image related to the entry point.

giuseppe replied:
the entire image is mounted as read only when running as a system container.  To have a writeable `/etc/passwd` we would need to bind mount a temporary file.

codificat commented on README_ANSIBLE_CONTAINER.md:
````diff
@@ -0,0 +1,29 @@
+* Running the ansible container:
+
+  ```sh
+  docker run -it --rm --privileged --net=host -v ~/.ssh:/root/.ssh -v /etc/ansible:/etc/ansible ansible
````
I'm wondering why this would need `--privileged` and `--net=host`? The image is expected to be run as a non privileged container (not even as UID 0)...

giuseppe replied:
I've dropped this part since running as a Docker container is already documented

codificat commented on system-container/config.json.template:
```diff
@@ -0,0 +1,229 @@
+			"CAP_NET_BIND_SERVICE",
+			"CAP_NET_BROADCAST",
+			"CAP_NET_ADMIN",
+			"CAP_NET_RAW"
```
Does the container really need these capabilities?

giuseppe replied:
I am going to drop these caps

giuseppe commented:
[test]

ashcrow commented:
aos-ci-test

codificat commented on Dockerfile:
```diff
@@ -16,6 +16,9 @@ LABEL name="openshift-ansible" \
 
+# Create a symlink to /opt/app-root/src so that files under /usr/share/ansible are accessible
+RUN mkdir -p /usr/share/ansible/ && ln -s /opt/app-root/src /usr/share/ansible/openshift-ansible
+
```
Why is this needed?

The "productized" container image built by `Dockerfile.rhel7` has the roles/playbooks under `/usr/share/ansible/openshift-ansible` because they are installed from the RPM, but I believe this is not needed in the upstream image. You can/should  use relative paths e.g. for `PLAYBOOK_FILE` and these would work the same in both cases, no?

giuseppe replied:
I've added it only to simplify working with the image as the same paths are valid also with `Dockerfile` and not having two different trees for the two images.  I can drop this patch if you prefer.

giuseppe replied:
I thought this would simplify how we work/debug the images as the same paths are valid.  I am going to drop this patch from the series for now.

codificat replied:
Now that we understand better the need for this, could we expand the comment a bit more? e.g. mention that the default `PLAYBOOK_FILE` in `config.json.template` must use a full path based on the RPM-based installation, and therefore this symlink is needed for this to work...

sdodson commented:
aos-ci-test

rhcarvalho commented on README_CONTAINER_IMAGE.md:
````diff
@@ -47,3 +47,17 @@ Here is a detailed explanation of the options used in the command above:
+  ```sh
+  atomic install --system --set INVENTORY_FILE=$(pwd)/inventory.origin openshift-ansible;
+  systemctl start openshift-ansible;
+  atomic uninstall openshift-ansible
````
This should probably be in a separate block? I imagine this is not part of how to run the container, but how to clean up.

rhcarvalho commented on system-container/run-system-container.sh:
```diff
@@ -1,3 +1,4 @@
 #!/bin/sh
 
+export ANSIBLE_LOG_PATH=/var/log/ansible.log
```
This commit/single line change could/should be squashed into the first commit of the PR: "installer: support running as a system container", the one that introduces this file.

rhcarvalho commented on system-container/tmpfiles.template:
```diff
@@ -1 +1,2 @@
 d    $VAR_LIB_OPENSHIFT_INSTALLER - - - - -
+f    $VAR_LOG_OPENSHIFT_LOG - - - - -
```
This commit can also we squashed, no need to introduce a two-line file in two separate commits.

rhcarvalho commented on system-container/config.json.template:
```diff
@@ -0,0 +1,221 @@
+                        "source": "$INVENTORY_FILE",
+                        "destination": "/etc/ansible/hosts",
+                        "options": [
+                                 "bind",
```
Looks like there's an extra space messing up the indentation here.

Here and a handful of other occurrences above inside of `"options"` blocks.

rhcarvalho commented:
@giuseppe @ashcrow, please, document the intent of this PR by filling in the PR description.

It's frustrating when we try to understand the history of a file / line and land into a PR with no description.

giuseppe commented on BUILD.md:
```diff
@@ -42,3 +42,17 @@ The progress of the build can be monitored with:
 
         oc describe imagestream openshift-ansible
+
+## Build the System Container
```
a system container doesn't run only on Atomic Host.  Is it fine to specify what the differences are and add a link to http://www.projectatomic.io/blog/2016/09/intro-to-system-containers/ ?

rhcarvalho replied:
Certainly, not being very familiar with system containers myself yet (except for your talk at DevConf :smile: ), that seems great! Thanks.

giuseppe commented on BUILD.md:
````diff
@@ -42,3 +42,17 @@ The progress of the build can be monitored with:
+```
+- Build the initial Docker image
+```
+docker build -t oainstaller
````
yes, going to change that and add a reference to the Docker build part

giuseppe commented:
@rhcarvalho I've addressed your comments.  Is it fine or would you like that I add more information?  The system container doesn't run only on Atomic Host, the only dependendencies are systemd, runc and the [atomic](https://github.com/projectatomic/atomic/) tool.

giuseppe commented on README_CONTAINER_IMAGE.md:
```diff
@@ -47,3 +47,26 @@ Here is a detailed explanation of the options used in the command above:
+
+Copy ssh public key of the host machine to master and nodes machines in the cluster.
+
+If the script needs additional files `/var/lib/openshift-installer` is bind mounted from the host (controllable with `VAR_LIB_OPENSHIFT_INSTALLER`)
```
I thought of this in case an user needs to deal with inventories that require external files.  An extra bind mount can be always useful :-)

giuseppe commented on README_CONTAINER_IMAGE.md:
````diff
@@ -47,3 +47,26 @@ Here is a detailed explanation of the options used in the command above:
+Run the ansible system container:
+
+```sh
+atomic install --system --set INVENTORY_FILE=$(pwd)/inventory.origin openshift/openshift-ansible
````
so the same image can be reused with different inventory files.  This defines the bind mount from the host file `$INVENTORY_FILE` to `/etc/ansible/hosts` in the container

rhcarvalho replied:
Sorry, this is confusing. The instructions tell users to copy the inventory to one place, but then reference the inventory with a different filename and at a different location. Would it be possible to clarify the example, perhaps make a clear separation / more than one example?

giuseppe replied:
I have added more information on what the $INVENTORY_FILE means, I hope this makes clearer how it can be used

rhcarvalho commented on README_CONTAINER_IMAGE.md:
```diff
@@ -47,3 +47,26 @@ Here is a detailed explanation of the options used in the command above:
+
+Copy ssh public key of the host machine to master and nodes machines in the cluster.
+
+If the inventory file needs additional files then it can use the path `/var/lib/openshift-installer` in the container as it is bind mounted from the host (controllable with `VAR_LIB_OPENSHIFT_INSTALLER`).
```
Re-reading this I understood it better, I think. Though I'm curious as to what is the corresponding path in the host?

giuseppe replied:
it should be configured with `--set VAR_LIB_OPENSHIFT_INSTALLER=/some/where`.  By default, if not configured it points to `/var/lib/openshift-installer`

codificat commented on BUILD.md:
```diff
@@ -42,3 +42,22 @@ The progress of the build can be monitored with:
+as described before.  The Docker image already contains all the
+required files to run as a system container.
+
+Once the Docker image is built, we can import it into the OSTree
```
Maybe replace the three instances of "Docker image" with "container image"?

codificat commented on README_CONTAINER_IMAGE.md:
```diff
@@ -47,3 +47,28 @@ Here is a detailed explanation of the options used in the command above:
 Additional usage information for images built from `playbook2image` like this one can be found in the [playbook2image examples](https://github.com/aweiteka/playbook2image/tree/master/examples).
+
+## Running Ansible as a System Container
+
```
Shouldn't this be "Running *openshift-ansible*..." instead of *Ansible*?

codificat commented on README_CONTAINER_IMAGE.md:
```diff
@@ -47,3 +47,28 @@ Here is a detailed explanation of the options used in the command above:
+
+Building the System Container: See the [BUILD.md](BUILD.md).
+
+Create `/etc/ansible` directory on the host machine and copy inventory file (hosts) into it.
```
I agree with Rodolfo that the part about the inventory is still confusing: you don't have to put your inventory in `/etc/ansible/hosts` in the host, right? it gets bind-mounted into that path in the container, but it can come from anywhere in the host, right?

giuseppe replied:
yes right, it can come from anywhere in the host.  I have dropped this line as this step is not really required.

codificat commented on system-container/root/exports/manifest.json:
```diff
@@ -0,0 +1,11 @@
+        "OPTS": "",
+        "VAR_LIB_OPENSHIFT_INSTALLER" : "/var/lib/openshift-installer",
+        "VAR_LOG_OPENSHIFT_LOG": "/var/log/ansible.log",
+        "PLAYBOOK_FILE": "/usr/share/ansible/openshift-ansible/playbooks/byo/config.yml",
```
As the symlink from `/usr/share/ansible` has been removed the `PLAYBOOK_FILE` should be a relative path (`playbooks/byo/config.yml`)

giuseppe replied:
we are using only one configuration file for the system container `config.json.template` while we have two Dockerfiles where it is possible to set the current working directory, duplicate the configuration file just for this difference doesn't seem ideal.  Users of Dockerfile will need to specify this argument as the default doesn't work.

Should we bring back the symlink?  It will make the two images work in the same way.

rhcarvalho replied:
The usage instructions today use relative paths (https://github.com/openshift/openshift-ansible/blob/master/README_CONTAINER_IMAGE.md), wouldn't it work in this case?

codificat replied:
Now that I understand this better I see more pros than cons, so ok for me to bring it back - I'd only ask that the comment around the symlink explains the where the requirement comes from a bit more (i.e. mention that the default `PLAYBOOK_FILE` must be specified with full path due to `cwd` in `config.json.template`...)

codificat replied:
> The usage instructions today use relative paths (https://github.com/openshift/openshift-ansible/blob/master/README_CONTAINER_IMAGE.md), wouldn't it work in this case?

took me a while to realize it wouldn't due to the difference: `cwd` in `config.json.template` vs `WORK_DIR` in `Dockerfile.rhel7`

rhcarvalho replied:
Aha, thanks @codificat, I haven't realized that before.

I'd still ask, since that file is a template, perhaps CWD could be part of the template?

I like to find simplicity, if we end up concluding that symlinks make it simpler, I'd favor it. But I'd consider the alternatives first -- update all image Dockerfiles to use the same paths and no need for symlinks; template the CWD in the config file, ...

Being able to provide a relative path makes the command-lines to interact with the images shorter, and relative to the repo root.

rhcarvalho commented on system-container/README.md:
```diff
@@ -0,0 +1,13 @@
+
+These files are needed to run the installer using an [Atomic System container](http://www.projectatomic.io/blog/2016/09/intro-to-system-containers/).
+
+* config.json.template - It is the configuration file for running the runC container.
```
This is a template of the config and not the config file itself ;)

Perhaps we could drop "runC" and say simply container. For someone who is not versed in all of the names and concepts, it can be confusing to see "system-container", "Atomic System Container" and "runC container" all in one place.

Suggestion:

```markdown
* `config.json.template` - Template of the configuration file used for running containers.
```

giuseppe commented:
I've added back the symlink.  Unfortunately we cannot template cwd (we would still need to give the correct value when running the container, similarly as we would do now for the path to the script).

I agree with using the same paths for the two images, we can take rid of these small differences like `cwd` and drop the symlink hack needed for the system container version.

giuseppe commented:
is there anything more blocking this PR?

giuseppe commented:
aos-ci-test

ashcrow commented:
```Internal Jenkins error during provisioning the jenkins slave.```

ashcrow commented:
aos-ci-test

rhcarvalho commented:
[merge]

sdodson commented:
aos-ci-test

ashcrow commented:
Excluder error. Retesting...

ashcrow commented:
aos-ci-test
